src/Workspaces/Core/Portable/Workspace/Workspace_Registration.cs

```diff
@@ -58,12 +55,24 @@ protected void UnregisterText(SourceTextContainer textContainer)
 
             var registration = GetWorkspaceRegistration(textContainer);
 
+            // guard us from being called with wrong text container.
             if (registration.Workspace == this)
             {
-                registration.SetWorkspaceAndRaiseEvents(null);
+                SetAndRaiseTextRegistrationChangedEvents(registration, newWorkspace: null);
             }
         }
 
+        private void SetAndRaiseTextRegistrationChangedEvents(WorkspaceRegistration registration, Workspace newWorkspace, [CallerMemberName] string eventName = null)
+        {
+            var oldWorkspace = registration.Workspace;
+            registration.SetWorkspace(newWorkspace);
+
+            // workspace change event is synchronous events. and it doesn't use workspace events queue since 
+            // WorkspaceRegistration is not associated with 1 specific workspace. workspace events queue is specific to 1 workspace
+            // using multiple queues that belong to multiple workspace will cause events to be out of order
+            registration.RaiseEvents(oldWorkspace, newWorkspace);
```

Comment:
i have 2 choice here. creating WorkspaceRegistration's own event queue to make it async events. or make these just synchronous events like I did here.

existing code was issue since not only one of them was async and the other is not, but because it used event queue of 1 workspace which can't guarantee event ordering of WorkspaceRegistration.WorkspaceChanged events.